### ISSUE TITLE
wrong result of SUM subquery results

### ISSUE THREAD

**Issue by Illyrix** [pingcap/tidb#17475]
### 1. Minimal reproduce step (Required)
init by following sql:
```sql
DROP TABLE IF EXISTS `table_int_float_varchar`, table_int_varchar;
CREATE TABLE `table_int_float_varchar` (
  `id_6` int(16) NOT NULL AUTO_INCREMENT,
  `col_int_6` int(16) DEFAULT NULL,
  `col_float_6` float DEFAULT NULL,
  `col_varchar_6` varchar(511) DEFAULT NULL,
  PRIMARY KEY (`id_6`)
);

INSERT INTO `table_int_float_varchar` VALUES (14,1,NULL,'0000-00-00 00:00:00'),(15,1,NULL,'NULL'),(16,1,1.5,NULL),(17,65535,NULL,' '),(18,-1,-0.1,'0'),(19,NULL,-0.1,'2020-02-02 02:02:00'),(20,1,0,NULL),(21,0,NULL,'NULL'),(22,-1,1.5,'0'),(23,-1,NULL,NULL),(24,1,NULL,'0');

CREATE TABLE `table_int_varchar` (
  `id_4` int(16) NOT NULL AUTO_INCREMENT,
  `col_int_4` int(16) DEFAULT NULL,
  `col_varchar_4` varchar(511) DEFAULT NULL,
  PRIMARY KEY (`id_4`)
);

INSERT INTO `table_int_varchar` VALUES (1,-1,'false'),(2,65535,NULL),(3,1,'-1'),(4,0,'NULL'),(5,1,'-0'),(6,0,'-0'),(7,65535,' '),(8,-1,'true'),(9,0,'0'),(10,NULL,'1'),(11,0,'-0'),(12,NULL,'false'),(13,NULL,'2020-02-02 02:02:00'),(14,1,NULL),(15,NULL,NULL),(16,NULL,NULL),(17,NULL,'0'),(18,1,NULL);


SELECT count(1) FROM table_int_varchar AS tmp1 LEFT JOIN table_int_float_varchar AS tmp2 ON (0e+00>=tmp2.id_6) IS NULL WHERE (tmp2.col_float_6 OR "") IS NULL;

SELECT sum(tmp_col) FROM (SELECT (tmp2.col_float_6 OR "") IS NULL AS tmp_col FROM (table_int_varchar AS tmp1 LEFT JOIN table_int_float_varchar AS tmp2 ON (0e+00>=tmp2.id_6) IS NULL)) AS tmp_table;
```

### 2. What did you expect to see? (Required)
execute in mysql:
```bash
mysql> SELECT count(1) FROM table_int_varchar AS tmp1 LEFT JOIN table_int_float_varchar AS tmp2 ON (0e+00>=tmp2.id_6) IS NULL WHERE (tmp2.col_float_6 OR "") IS NULL;
+----------+
| count(1) |
+----------+
|       18 |
+----------+
1 row in set (0.05 sec)

mysql> SELECT sum(tmp_col) FROM (SELECT (tmp2.col_float_6 OR "") IS NULL AS tmp_col FROM (table_int_varchar AS tmp1 LEFT JOIN table_int_float_varchar AS tmp2 ON (0e+00>=tmp2.id_6) IS NULL)) AS tmp_table;
+--------------+
| sum(tmp_col) |
+--------------+
|           18 |
+--------------+
1 row in set (0.05 sec)
```
the two SQLs are equivalent, so the results should be same
### 3. What did you see instead (Required)
in tidb:
```bash
mysql> SELECT count(1) FROM table_int_varchar AS tmp1 LEFT JOIN table_int_float_varchar AS tmp2 ON (0e+00>=tmp2.id_6) IS NULL WHERE (tmp2.col_float_6 OR "") IS NULL;
+----------+
| count(1) |
+----------+
|        0 |
+----------+
1 row in set (0.00 sec)

mysql> SELECT sum(tmp_col) FROM (SELECT (tmp2.col_float_6 OR "") IS NULL AS tmp_col FROM (table_int_varchar AS tmp1 LEFT JOIN table_int_float_varchar AS tmp2 ON (0e+00>=tmp2.id_6) IS NULL)) AS tmp_table;
+--------------+
| sum(tmp_col) |
+--------------+
|           18 |
+--------------+
1 row in set (0.00 sec)
```
### 4. Affected version (Required)

```
Release Version: v4.0.0-beta.2-499-gc59cddcc0
Edition: Community
Git Commit Hash: c59cddcc01e46adf4520d161f10a8e19bb58b785
Git Branch: master
UTC Build Time: 2020-05-22 07:35:25
GoVersion: go1.13.5
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

**Comment by zz-jason** [comment]
@winoros PTAL

**Comment by winoros** [comment]
The `col_float_6 or ""` is converted to `col_float_6 is true or ""`. So we treat the expression as null-rejected and convert the outer join to inner join.

**Comment by francis0407** [comment]
This is fixed by #19621
